# 001 - Sentence Embeddings

Concept: Rep sentences as numerical vectors using transformer model

Question: how embedding happens, how it looks like?

In [11]:
%pip install -q sentence-transformers

In [12]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('all-MiniLM-L6-v2')

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

In [13]:
sentences =[
    "Lakshay is a Data Science and Engineering Major.",
    "He wants to learn swimming.",
    "He has an offer in hand.",
]

In [14]:
embeddings = model.encode(sentences)

print("Complete Matrix Shape:", embeddings.shape)

for sentence, embedding in zip(sentences, embeddings):
    print("\nSentence:", sentence)
    print("Embedding shape:", embedding.shape)
    print("First five values:",embedding[:5]) 

Complete Matrix Shape: (3, 384)

Sentence: Lakshay is a Data Science and Engineering Major.
Embedding shape: (384,)
First five values: [ 0.00204887 -0.05346763 -0.02676995 -0.00756089  0.00948199]

Sentence: He wants to learn swimming.
Embedding shape: (384,)
First five values: [ 0.066445    0.07940186  0.00026348 -0.00273268 -0.07613984]

Sentence: He has an offer in hand.
Embedding shape: (384,)
First five values: [-0.05321697  0.05134412 -0.01160448  0.03473996 -0.08831972]


In [15]:
sentences.append("Machine Learning is what he wants to do.")

new_embeddings = model.encode(sentences)

print("old shape:", embeddings.shape)
print("new shape:", new_embeddings.shape)

old shape: (3, 384)
new shape: (4, 384)


## Observations

- the matrix shape first was (3,384)
- when i added one sentences it became (4,384)
- one interesting this was in the original sentence database when i forgot to put a comma after the last sentence the shape showed up as (1,384) instead of (3,384)

- one very interesting this is i wondered why the word "he" has different embedding value for sentence 2 and 3 when then word is same...
so apparently the numbers are not the embedding of the word he, they are the first five dimmensions of the embedding for the entire sentence.

"He wants to learn swimming" - one 384-d sentence vector
"He has an offer in hand" - another 384-d sentence vector

- the model processes each word together , through attention each word is encoded wrt to its surrounding (context capture).
he in first sentence is connected with "want", "learn" and "swimming".
he in the second sentence is connected with "offer" and "hand"

## Concept

Sentence Transformers convert a sentence into a fixed-size numerical vector.

Tokens → contextual token representations → pooling → 384-dimensional sentence embedding

This differs from older static embedding methods:

- **Word2Vec:** A word usually has the same stored vector wherever it appears.
- **Transformer:** A token’s representation changes based on the surrounding context.


